src/types.rs

```diff
@@ -481,6 +492,398 @@ impl ZipFileData {
+                self.uncompressed_size.try_into().unwrap(),
+            )
+        };
+
+        let mut extra_field_length = self.extra_field_len();
+        if self.large_file {
+            /* TODO: magic number */
+            extra_field_length += 20;
+        }
+        if extra_field_length + self.central_extra_field_len() > u16::MAX as usize {
+            return Err(ZipError::InvalidArchive("Extra data field is too large"));
+        }
+        let extra_field_length: u16 = extra_field_length.try_into().unwrap();
+
+        Ok(ZipLocalEntryBlock {
+            magic: spec::LOCAL_FILE_HEADER_SIGNATURE,
+            version_made_by: self.version_needed(),
+            flags: self.flags(),
+            #[allow(deprecated)]
+            compression_method: self.compression_method.to_u16(),
+            last_mod_time: self.last_modified_time.timepart(),
+            last_mod_date: self.last_modified_time.datepart(),
+            crc32: self.crc32,
+            compressed_size,
+            uncompressed_size,
+            file_name_length: self.file_name_raw.len().try_into().unwrap(),
+            extra_field_length,
+        })
+    }
+
+    pub(crate) fn block(&self, zip64_extra_field_length: u16) -> ZipEntryBlock {
+        let extra_field_len: u16 = self.extra_field_len().try_into().unwrap();
+        let central_extra_field_len: u16 = self.central_extra_field_len().try_into().unwrap();
+        ZipEntryBlock {
+            magic: spec::CENTRAL_DIRECTORY_HEADER_SIGNATURE,
+            version_made_by: (self.system as u16) << 8 | (self.version_made_by as u16),
+            version_to_extract: self.version_needed(),
+            flags: self.flags(),
+            #[allow(deprecated)]
+            compression_method: self.compression_method.to_u16(),
+            last_mod_time: self.last_modified_time.timepart(),
+            last_mod_date: self.last_modified_time.datepart(),
+            crc32: self.crc32,
+            compressed_size: self
+                .compressed_size
+                .min(spec::ZIP64_BYTES_THR)
+                .try_into()
+                .unwrap(),
+            uncompressed_size: self
+                .uncompressed_size
+                .min(spec::ZIP64_BYTES_THR)
+                .try_into()
+                .unwrap(),
+            file_name_length: self.file_name_raw.len().try_into().unwrap(),
+            extra_field_length: zip64_extra_field_length
+                + extra_field_len
+                + central_extra_field_len,
+            file_comment_length: self.file_comment.as_bytes().len().try_into().unwrap(),
+            disk_number: 0,
+            internal_file_attributes: 0,
+            external_file_attributes: self.external_attributes,
+            offset: self
+                .header_start
+                .min(spec::ZIP64_BYTES_THR)
+                .try_into()
+                .unwrap(),
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(packed)]
+pub(crate) struct ZipEntryBlock {
```

Comment:
Rename this to `ZipCentralEntryBlock` to clarify the difference vs `ZipLocalEntryBlock`. Also, can these two structs have the accessors and mutators for fields they both include, factored out into a trait? `ZipEntryBlock` will be available for the trait name.

Reply:
I think this is a legitimately good idea, but I've been stumped on where exactly to put it. We really don't access the fields of these `Block` structs very much except to coalesce them into fields of `ZipFileData`, and I would prefer to maintain that control flow (and make all these `Block` structs `pub(crate)`) instead of adding more helpers to access those fields. So to answer your question, I think it does not eliminate very much code to factor out an accessor trait here, because we are only using these `Block` structs temporarily as part of generating the `ZipFileData`.